python/spglib/spglib.py

```diff
@@ -97,6 +97,20 @@ class SpglibError:
 spglib_error = SpglibError()
 
 
+class MagneticSpaceGroupType(TypedDict):
```

Comment:
Have you experimented with combining `TypedDict` and `@dataclass`? Ideally we should start making the latter interface available since it has better type-hinting and navigability. We can't switch it to only `dataclass` because of backwards compatibility.

If they can't be combined, I can write a quick `Mapping` interface to provide backwards compatibility with warning emission to switch over.

Reply:
> Have you experimented with combining TypedDict and @dataclass? 

No, I haven't. I've picked TypedDict here just from backward compatibility.

Do you mean `collections.abc.Mapping` for the interface?

Reply:
> Do you mean `collections.abc.Mapping` for the interface?

Yes, but that's only as a backup if subclassing `TypedDict` + `@dataclass` decorator doesn't work. Both of those seem to use PEP526. My concern is that both of them may overload the constructor which will clash

Reply:
I did some testing and just combining `TypedDict` and `@dataclass` did not work (technically it compiled, but it lost all attributes of `@dataclass`). I've tried it with `attrs` as well, but there a clear error was produced. Probably `pydantic` would have a similar issue. Here is a test script for `Mapping` + `@dataclass`

```python
from __future__ import annotations

import dataclasses
from collections.abc import Mapping
from typing import Any, Iterator


@dataclasses.dataclass
class A(Mapping[str, Any]):
    a: int
    b: str

    def __getitem__(self, key: str) -> Any:
        return dataclasses.asdict(self)[key]

    def __len__(self) -> int:
        return len(dataclasses.fields(A))

    def __iter__(self) -> Iterator[str]:
        return iter(dataclasses.asdict(self))
```

The mapping is not typed as `TypedDict` unfortunately, but at least it does have all of the read-only interface of a dict with this.

Reply:
My suggestion here is `dataclass` in order to start providing attribute interface (`A.a`), and combine it with subclassing with `Mapping` in order to maintain dict interface (`A['a']`). The latter I suggest to start deprecating because it is:
- less supported with type-hint
- less maintainable and recommended
- harder to document

Indeed `TypedDict` does offer some similar functionalities with `dataclass`, but it only offers type-hint improvements. `dataclass`/`attrs`/`pydantic` are the more recommended class containers.

There are also `pandas` and `xarray`, but these are a bit overkill for the datastructure that is meant to be provided here, and you can transform one to another by piping `dataclass`/`attrs`/`pydantic` -> `dict` (there are built-in interfaces for this) -> `pandas`/`xarray` (we can also provide simple interface for these but we probably need to consult the community of some naming conventions for units, matrix transformations, etc.)

Reply:
I totally agree that we should finally migrate to `dataclass`. But, I'm worry about using such a uncommon practice to combine Mapping and dataclass. I feel it is better to use dataclass all at once.

Reply:
> I feel it is better to use dataclass all at once

That would be hard for the users because it would break the API.

> But, I'm worry about using such a uncommon practice to combine Mapping and dataclass

It's actually not a bad approach. `Mapping` does not define `__init__`, and the only ones that are [added](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) do not clash with `dataclass`, except for `__eq__` and `__ne__`, which are easy to fix.

As for the commonality, there is a package that offers similar syntax, [`box`](https://github.com/cdgriffith/Box/), a few user-side projects that do the same (`postopus` from my institute), and of course `pandas`/`xarray` which have a more complex system. This approach, is even simpler though because we explicitly do not want the user to edit these fields, so `Mapping` is sufficient instead of `MutableMapping` (`dict`-view vs `dict`).

Reply:
Thanks for your detail explanation. The [list of Mapping methods](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) is convincing for me.
Let's use `dataclass` with `Mapping` for `SpaceGroupType` and `MagneticSpaceGroupType`. The introduction for other APIs would be in another PR.